### ISSUE TITLE
Unnecessary var assign?

### ISSUE THREAD

**Issue by edsonmedina** [scrutinizer-ci/scrutinizer#289]
I'm getting this issue reported when assigning a variable to null (to free the memory, as opposed to unset() which merely destroys the reference).

see [Issue](https://scrutinizer-ci.com/g/edsonmedina/php_testability/issues/master/files/src/HTMLReport.php?orderField=path&order=asc&issueId=a1ae17f4-0b45-4930-a1a8-967fd462b04d) (Id: dead_assignment.unnecessary_var_assign)

**Comment by edsonmedina** [comment]
That's an oversimplification.

AFAIK unset() marks the variable for deletion, but the memory won't be freed until the next garbage collecting cycle, while setting to null frees the memory immediately (which costs CPU cycles, so it's slower). It's a trade-off.

In my specific case, saving memory is more important that running fast.

**Comment by stof** [comment]
@edsonmedina you are wrong. ``unset`` deletes the variable, which decrements the refcount of the zval, exactly the same than when assigning null to the variable (where the previous zval is not reference anymore). Waiting for the next GC cycle is needed only in case there is a circular ref in your object graph, and there won't be any difference when assigning ``null``.

The only difference when assigning ``null`` is that you consume memory for an unnecessary variable referencing a ``null`` zval.

**Comment by edsonmedina** [comment]
So you're saying the GC will fire immediately with unset() (ie: even without going out of scope)?
ie: 

    function foo() 
    {
        // insert memory hungry operation here
    
        unset ($huge_var); // GC will collect here?
     
        // insert another memory hungry operation here
    }

If that's the case, do you have any references to back that up?

**Comment by stof** [comment]
No, I'm saying that in many case, PHP variables are **not** destroyed by the GC but by the reference counting mechanism (the GC is only responsible for cyclic object graphs for which counting references is not enough).

And deleting the variable or assigning a different value in this variable makes strictly no difference for what happens to the previous value. In both case, the effect is exactly the same: decrementing the refcount of the zval by 1 because it removes the reference to the value which was hold by the variable.

In both case, you are removing the reference. Assigning null to a variable does not free memory at all. There is no way to free memory from PHP code as memory is entirely managed internally (all you can do is know how references are preventing the memory to be freed)
